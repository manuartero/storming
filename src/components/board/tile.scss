$available-tile: rgba(195, 214, 191, 0.5);
$forbidden-tile: rgba(190, 88, 88, 0.5);
$tile-border: black;

.tile {
  display: flex;
  position: relative;
  width: 120px;
  height: 120px;
  margin: 0;
  justify-content: center;
  align-items: center;

  &__id {
    margin: auto;

  }

  &__terrain {
    position: absolute;
    stroke: #5f5858;
    stroke-width: 3px;

    &--forest {
      top: 15%;
      left: 15%;

      width: 80px;
      height: 80px
    }

    &--lake {
      top: 5%;
      left: 5%;
      width: 110px;
      height: 110px
    }
  }


  &__hexagon {
    position: absolute;
    width: 120px; // needed when changed from CRA to Vite
    height: 120px; // needed when changed from CRA to Vite

    path {

      fill: none;
      stroke: $tile-border;
      stroke-width: 6px;
    }

    &--selected {
      path {
        fill: rgba(255, 248, 174, 0.179);
      }
    }

    &--available {
      path {
        fill: $available-tile;
      }
    }

    &--forbidden {
      path {
        fill: $forbidden-tile;
      }
    }
  }


  .pointer-area {

    /* inner area */
    &::before {
      content: '';
      position: absolute;
      top: 3%;
      left: 3%;
      width: 94%;
      height: 94%;

      cursor: pointer;

      clip-path: polygon(50% 0%, 100% 25%, 100% 75%, 50% 100%, 0% 75%, 0% 25%);

      // debug
      // background-color: rgba(255, 0, 0, 0.292);
    }
  }

  // center
  &__content {


    display: flex;

  }
}
