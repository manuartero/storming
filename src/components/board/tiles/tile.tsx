import c from "classnames";
import { coordinates } from "models/tiles";
import { tileAssets } from "./assets";

import styles from "./tile.module.css";
import { usePieceOffset } from "./use-piece-offset";

type Props = React.PropsWithChildren<{
  id: TileID;
  status?: TileStatus;
  terrain?: TerrainType;
  building?: BuildingType;
  owner?: PlayerType;
  activePlayer?: PlayerType;
  onClick: (tileID: Coordinates) => void;
}>;

export function Tile({
  id,
  terrain,
  building,
  owner,
  activePlayer,
  status,
  children,
  onClick,
}: Props) {
  /* debug: turn on debug ID */
  const debugTileID = false;

  const pieceStyle = usePieceOffset(children);

  return (
    <button
      key={`tile(${id})`}
      className={c(
        styles.tile,
        status && styles[status],
        owner && styles[owner],
        owner && styles.controlled,
        status === "available" && activePlayer && styles[activePlayer]
      )}
      aria-label={`tile ${id}`}
      aria-disabled={status === "forbidden"}
      type="button"
      onClick={() => onClick(coordinates(id))}
    >
      <div className={styles.strokeLayer} aria-hidden="true" />

      <div className={styles.innerLayer} aria-hidden="true">
        {terrain && <Terrain variant={terrain} />}
        {debugTileID && <span className={styles.tileId}>{id}</span>}
      </div>

      {building && owner && <Building variant={building} owner={owner} />}
      {children && (
        <div className={c(styles.piece)} style={pieceStyle}>
          {children}
        </div>
      )}
    </button>
  );
}

function Terrain({ variant }: { variant: TerrainType }) {
  const icon = tileAssets.terrain[variant];

  return (
    <div
      role="img"
      className={c(styles.terrain)}
      aria-roledescription="game terrain"
      aria-label={`terrain ${variant}`}
      style={{ backgroundImage: `url(${icon})` }}
    />
  );
}

function Building({
  variant,
  owner,
}: {
  variant: BuildingType;
  owner: PlayerType;
}) {
  const icon = tileAssets.buildings[variant][owner];

  return (
    <div
      role="img"
      className={c(styles.building)}
      aria-roledescription="game building"
      aria-label={`${owner} ${variant}`}
      style={{ backgroundImage: `url(${icon})` }}
    />
  );
}
