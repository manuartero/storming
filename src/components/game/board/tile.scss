@use '/src/styles/colors.scss';

$forbidden-tile: rgba(190, 88, 88, 0.5);

$horizontal-padding: 5px;

.tile {
  position: relative;
  width: 120px;
  height: 120px;
  margin: 0;
  padding: 0 $horizontal-padding;

  display: flex;
  justify-content: center;
  align-items: center;

  /* debug: turn on at tile.tsx */
  &__id {
    position: absolute;
    font-size: larger;
    background-color: white;
    z-index: 1;
  }

  &__content {}

  &__terrain {
    position: absolute;
    stroke: #5f5858;
    stroke-width: 3px;

    &--forest {
      top: 15%;
      left: 15%;

      width: 80px;
      height: 80px
    }

    &--lake {
      top: 5%;
      left: 5%;
      width: 110px;
      height: 110px
    }
  }

  &__building {
    width: 90px;
    height: 90px;

    position: absolute;
    bottom: 5px; // building in the bottom

    background-size: 100% 100%;
    background-repeat: no-repeat;
    background-position: center;

    &--player {
      background-image: url('./assets/castle--player.png');
    }

    &--enemy1 {
      background-image: url('./assets/castle--enemy1.png');
    }

    &--enemy2 {
      background-image: url('./assets/castle--enemy2.png');
    }

    &--enemy3 {
      background-image: url('./assets/castle--enemy3.png');
    }
  }

  &__hexagon {
    position: absolute;

    path {
      fill: none;
    }

    &--selected {
      path {
        fill: colors.$selected-tile;
        stroke: colors.$dark-grey;
        stroke-width: 3px;
      }

      &-player {
        path {
          stroke: colors.$player-primary;
        }
      }

      &-enemy1 {
        path {
          stroke: colors.$enemy1-primary;
        }
      }

      &-enemy2 {
        path {
          stroke: colors.$enemy2-primary;
        }
      }

      &-enemy3 {
        path {
          stroke: colors.$enemy3-primary;
        }
      }
    }

    &--available {
      &-player {
        path {
          stroke: colors.$player-primary;
        }
      }

      &-enemy1 {
        path {
          stroke: colors.$enemy1-primary;
        }
      }

      &-enemy2 {
        path {
          stroke: colors.$enemy2-primary;
        }
      }

      &-enemy3 {
        path {
          stroke: colors.$enemy3-primary;
        }
      }
    }

    &--forbidden {
      path {
        fill: $forbidden-tile;
      }
    }

    &__clickable-area {
      clip-path: polygon(50% 0%, 100% 25%, 100% 75%, 50% 100%, 0% 75%, 0% 25%);

      // debug
      // background-color: rgba(255, 0, 0, 0.292);

      // FIXME: this is absolutly manual, there should be a better way of solving this.
      position: absolute;
      top: 5%;
      left: 5%;
      width: 90%;
      height: 90%;
      // ---
    }
  }

}
