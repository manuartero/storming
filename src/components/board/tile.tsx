import c from "classnames";
import { useState } from "react";
import { ReactComponent as ForestSvg } from "./assets/forest.svg";
import { ReactComponent as HexagonSvg } from "./assets/hexagon.svg";
import { ReactComponent as LakeSvg } from "./assets/lake.svg";
import { ReactComponent as MountainSvg } from "./assets/mountain.svg";
import tileId from "./tile-id";

import "./tile.scss";

interface Props {
  id: TileIdStr;
  terrain?: "field" | "mountain" | "lake" | "forest";
  status?: "idle" | "available" | "forbidden" | "selected";
  onClick?: (tileID: TileID) => void;
  children?: React.ReactNode;
}

function Tile({ id, terrain, status, children, onClick }: Props): JSX.Element {
  // console.debug(`render <Tile id="${id}" />`)
  const tileID = tileId(id);

  return (
    <div className="tile">
      {terrain === "mountain" && (
        <MountainSvg className={c("tile__terrain", "tile__terrain-mountain")} />
      )}
      {terrain === "forest" && (
        <ForestSvg className={c("tile__terrain", "tile__terrain--forest")} />
      )}
      {terrain === "lake" && (
        <LakeSvg className={c("tile__terrain", "tile__terrain--lake")} />
      )}

      <HexagonSvg
        className={c(
          "tile__hexagon",
          status === "available" && "tile__hexagon--available",
          status === "forbidden" && "tile__hexagon--forbidden",
          status === "selected" && "tile__hexagon--selected"
        )}
      />

      <div
        className="pointer-area"
        onClick={() => {
          onClick && onClick(tileID);
        }}
      />
      {children && <div className="tile__content">{children}</div>}
    </div>
  );
}

export default Tile;
