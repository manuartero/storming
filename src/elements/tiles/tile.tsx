import c from "classnames";
import { coordinates } from "models/tiles";
import { tileAssets } from "./assets";
import { usePieceOffset } from "./use-piece-offset";

import styles from "./tile.module.css";

type _BaseProps = React.PropsWithChildren<{
  status?: TileStatus;
  terrain?: TerrainType;
  building?: BuildingType;
  owner?: PlayerType;
  activePlayer?: PlayerType;
  disableChildrenOffset?: boolean;
}>;

type _TileProps = _BaseProps & {
  id: TileID;
  onClick: (tileID: Coordinates) => void;
};

type _FakeTile = _BaseProps & {
  id: null;
  onClick?: undefined;
};

type Props = _TileProps | _FakeTile;

export function Tile({
  id,
  terrain,
  building,
  owner,
  activePlayer,
  status,
  children,
  disableChildrenOffset = false,
  onClick,
}: Props) {
  /* debug: turn on debug ID */
  const debugTileID = false;

  const pieceStyle = usePieceOffset({ children, disableChildrenOffset });

  return (
    <button
      className={c(
        styles.tile,
        status && styles[status],
        owner && styles[owner],
        owner && styles.controlled,
        status === "available" && activePlayer && styles[activePlayer]
      )}
      aria-label={`tile ${id}`}
      aria-disabled={status === "forbidden"}
      type="button"
      onClick={id ? () => onClick(coordinates(id)) : undefined}
    >
      <div className={styles.strokeLayer} aria-hidden="true" />

      <div className={styles.innerLayer} aria-hidden="true">
        {terrain && <Terrain variant={terrain} />}
        {debugTileID && <span className={styles.tileId}>{id}</span>}
      </div>

      {building && owner && <Building variant={building} owner={owner} />}
      {children && (
        <div className={c(styles.piece)} style={pieceStyle}>
          {children}
        </div>
      )}
    </button>
  );
}

function Terrain({ variant }: { variant: TerrainType }) {
  const icon = tileAssets.terrain[variant];

  return (
    <div
      role="img"
      className={c(styles.terrain)}
      aria-roledescription="game terrain"
      aria-label={`terrain ${variant}`}
      style={{ backgroundImage: `url(${icon})` }}
    />
  );
}

function Building({
  variant,
  owner,
}: {
  variant: BuildingType;
  owner: PlayerType;
}) {
  const icon = tileAssets.buildings[variant][owner];

  return (
    <div
      role="img"
      className={c(styles.building)}
      aria-roledescription="game building"
      aria-label={`${owner} ${variant}`}
      style={{ backgroundImage: `url(${icon})` }}
    />
  );
}
