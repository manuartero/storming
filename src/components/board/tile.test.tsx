import { render } from "@testing-library/react";
import { Tile } from "./tile";

describe("<Tile />", () => {
  test("render: match snapshot", () => {
    const { asFragment } = render(<Tile id="-1,0" onClick={jest.fn()} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
