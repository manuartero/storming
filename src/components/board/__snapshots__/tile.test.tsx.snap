// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`<Tile /> render: button role 1`] = `
<button
  aria-disabled="false"
  aria-label="tile -1,0"
  class="tile"
  type="button"
>
  <svg
    class="tile__hexagon"
  >
    svg-mock
  </svg>
  <div
    class="tile__hexagon__clickable-area clickable"
  />
</button>
`;
